asm/instruction.go

```diff
@@ -935,14 +935,19 @@ func (iter *InstructionIterator) Next() bool {
 type bpfRegisters uint8
 
 func newBPFRegisters(dst, src Register, bo binary.ByteOrder) (bpfRegisters, error) {
+	buf := make([]byte, 2)
+	val := uint16(dst&0x0F)<<8 | uint16(src&0x0F)
 	switch bo {
 	case binary.LittleEndian:
```

Comment:
why not replacing this whole switch with `bo.PutUint16` directly ?

Reply:
That's the annoying bit: calling an interface method cannot be inlined, plus it allocates. This function is called a ton, so it starts adding up.

We have some amount of discovery to do, because now binary.NativeEndian is a thing, we can probably eliminate these kind of switches/branches and replace them with `binary.NativeEndian.PutUint16`, since marshaling bpf insns is usually done before loading into the host kernel, and we'd almost always do that in native endianness. 

`Instruction.Marshal()` is exported API, though, so we should have a look at public callers on Sourcegraph to see if anyone depends on marshaling asm.Instructions in arbitrary endiannesses (e.g. for converting ELFs from LE to BE).

Reply:
ah yeah, that's too bad

Reply:
> `Instruction.Marshal()` is exported API, though, so we should have a look at public callers on Sourcegraph to see if anyone depends on marshaling asm.Instructions in arbitrary endiannesses

I'd rather not get into the habit of hardcoding NativeEndian, it leads to not thinking about endianness at callsites and that tends to cause subtle bugs. Compared to that the type switch isn't bad at all imo.